{% extends "base.html" %}

{% block content %}

<center>
<h1 style="font-variant:small-caps">MutationExplorer</h1>
</center>

<p>
  The server lets users mutate proteins in
  various ways and study mutation effects in 3D.
  The only required inputs are a structural model and a list of mutations of
  interest.
  Alternatively, the user can upload human SNPs in the VCF format.
</p>
  
<h4>Main steps</h4>
<p> <b>Either:</b>
  <ul><li> upload human SNPs in VCF format</li></ul>
  <b>Or:</b>
  <ul>
    <li> upload structure </li>
    <li> define mutations </li>
    <li> mutate again and again ... </li>
  </ul>
</p>

<h4> Structure upload </h4>

<p>
  Structures can be uploaded in three different ways:
  <ul>
    <li> PDB ID </li>
    <li> AlphaFold ID </li>
    <li> Upload local PDB file </li>
  </ul>
</p>

<h4> Define mutations </h4>

<p>
  Mutations can be defined in three different ways:
  <ul>
    <li> Key or paste into GUI, using a simple syntax:</li>
    <li> Upload alignment in ClustalW format </li>
    <li> Upload target sequence either by pasting the sequence, uploading a FASTA file or specifying a Uniprot ID </li>
  </ul>
</p>

<h4> Mutation syntax </h4>
<ul>
  <li> A comma separated list of CHAIN ':' RESIDUE-ID TARGET-AA-TYPE</li>
  <li> Example: 'A:23G,B:123A'</li>
  <li> This mutates the residue in chain A with the PDB residue ID 23 to Glycine and in chain B the residue with residue ID 123 to Alanine.</li> 
</ul>



<h4> Explorations </h4>
<p> You can select a model and further mutate this using the mutation window. </p>

<h4> Expected time requirements </h4>

<p>
  Actual times depend on the size of the system and may vary significantly.
  <ul>
    <li> Initial basic minimization: 1 min </li>
    <li> Mutation and basic minimization: 2 min  </li>
    <li> VCF processing:  5 min  </li>
  </ul>
</p>

<p> Details can be found in the FAQs </p>

<br>
<br>
<br>

{% endblock %}
