{% extends "base.html" %}

{% block head %}
<style>
    .bld {
        font-weight: 600;
    }

    .true-list {
        padding-left: revert !important;
    }

    .true-list li {
        list-style-type: disc !important;
    }

    .mut-ex {
        font-variant: small-caps;
    }

    .collapsible-header .mut-ex {
        margin: 0 0.4rem;
    }

    .collapsible-body {
        text-align: justify;
    }
</style>
{% endblock %}


{% block content %}

<center>
    <h3>Tutorials for <span style="font-variant:small-caps">MutationExplorer</span> </h3>


    <!--

<p>
  The server lets users mutate proteins in
  various ways and study mutation effects in 3D.
  The only required inputs are a structural model and a list of mutations of
  interest.
  Alternatively, the user can upload human SNPs in the VCF format.
  <br>
  You can click on the       <span class="center-align toggle-card"> <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="Further information" style="margin:5px;">i</button> </span> buttons at each field in the forms to get further information.
  <div id="modal3" class="modal">
    <div class="modal-content">
      This button has no function in the tutorial.
      <b>Happy mutating!</b>
      <p> </p>
    </div>
    <div class="modal-footer">
      <a href="#!" class="modal-close waves-effect waves-green btn-flat">Close</a>
    </div>
  </div>
</p>
-->

    <!-- <h5>General philosophy and key elements of the <span class="mut-ex">MutationExplorer</span></h5>
    <center>
        <div><img src="static/img/server2b.png" width="62%;"></div>
    </center>
    <ul class="collapsible">

        <li>
            <div class="collapsible-header grey lighten-4">
                <span class="mut-ex">MutationExplorer</span> - exploring the effect of mutations on the structure
            </div>
            <div class="collapsible-body"><span>
                    The <span class="mut-ex">MutationExplorer</span> maps variants onto the protein 3D structure which
                    allows to
                    interactively explore the effects of mutations with respect to stability and function. Often wwPDB
                    structures
                    contains mutations and can not intuitively be mapped to its native sequence, sequences of structures
                    changed
                    to close or remote homologs or structures designed interactive while inspecting the structure. With
                    the <span class="mut-ex">MutationExplorer</span> this is now easily possible and even visualises the
                    effect of the
                    variant on the protein structures stability.
                </span></div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Energy minimization - the key to high quality
            </div>
            <div class="collapsible-body"><span>
                    <p>
                        The minimization is crucial for the quality of the outcome. The better the structure is
                        energetically
                        minimized, the more reliable the results will be. <br>
                        We continuously pre-minimize structures from the PDB using the commandline described below.
                        If you specify a PDB ID that is contained in our database, the pre-minimized structure is used.
                        <br><br>
                        Otherwise we offer two modes of side-chain minimization, a short and a long one. The short one
                        should take a
                        few minutes, depending on the size of the protein. The long one may take some hours for large
                        proteins.
                        It is strongly recommended to use the long minimization, despite the significant waiting period.
                    </p>
                    <b> AlphaFold and your own models</b>
                    <p>These are not pre-minimized and you should follow the steps below. The AlphaFold models should be
                        minimized, but the minimization has to be performed with the same energy functions than we use
                        in the <span class="mut-ex">MutationExplorer</span>, namely Rosetta. </p>
                    <strong> Do it yourself!</strong>
                    <p> The minimizations we offer as 'long' and 'short' are both only side-chain optimizations.
                        However, we recommend you to perform the following minimization. It is optimizing also the
                        backbone in a
                        limited way (due to '-relax:constrain_relax_to_start_coords'). A free backbone minimization
                        might be leading
                        to problems for e.g. membrane proteins. </p>
                    <div class="codebox">
                        <code>
                PATH_TO_ROSETTA/relax.static.linuxgccrelease   -relax:fast true -relax:cartesian true -score:weights ref2015_cart -use_input_sc -optimization::default_max_cycles 200 -linmem_ig 10 -relax:constrain_relax_to_start_coords -ex1 -ex2  -nstruct 20 -in:file:s  INPUT.pdb -out:pdb -out:prefix  OUTDIR/
              </code>
                    </div>
                    <p>
                        This commandline will create 20 models, from which you should select the one with the lowest
                        energy. Each
                        PDB contains a line starting with 'pose'. In this line you want the last value, which is the
                        total energy.
                    </p>

                    <b>ROSIE server</b>

                    <p>
                        There is a way to perform the structure minimization without Rosetta installation: using the
                        ROSIE 2 server
                        (https://r2.graylab.jhu.edu/). It requires that you have a GitHub account with your academic
                        email address
                        linked and verified. Once you are logged in, you can select ‘Rosetta’ from Rosie apps list.
                        There, you need
                        to upload your PDB structure and paste the following into the ‘Rosetta flags file’ field:
                    </p>
                    <div class="codebox">
                        <code>
		  -in:file:s input.pdb<br>
		  -in:file:fullatom<br>
		  -constrain_relax_to_start_coords<br>
		  -relax:ramp_constraints false<br>
		  -nstruct 6<br>
		  -out:suffix _Relaxed<br>		  
		  #rosie type serial<br>
		  #rosie tasks 6
	      </code>
                    </div>
                    <p>
                        The job will run for about 2-4 hours, depending on the protein size. Once it is done, you can go
                        to ‘job
                        files’ and find ‘score_Relaxed.sc’. You can open it on the website, if you do so, you need to
                        select ‘raw’
                        in the ‘view as’ options. In the ‘total score’ column, find the row with the lowest energy score
                        and look up
                        the name in the ‘description’ column. Then download the ‘input_relaxed’ .pdb file with the
                        respective
                        number. This is the file you are going to use as an input to the MutationExplorer.
                    </p>

                    <p>
                        <b>Database of pre-minimized structures from the Protein Databank (PDB)</b><br><br>
                        Since the minimization is essential for correctly highlighting the energetic impact of
                        mutations, but takes
                        no small amount of time to perform, we calculated a database of pre-minimized structures from
                        the PDB.
                        Currently our database contains 45,000 models. You will be notified if your PDB ID was found in
                        our database
                        or has to be calculated.<br>
                    </p>
                </span></div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-4">
                RaSP - bringing exploration to full bloom
            </div>
            <div class="collapsible-body"><span>
                    <p>
                        RaSP is a new deep-learning-based tool that rapidly
                        estimates protein stability changes. RaSP predictions
                        strongly correlate to scores from Rosetta calculations
                        which demand longer compute times. <br><br>
                        With RaSP, <span style="font-variant:small-caps">MutationExplorer</span>
                        presents the user with a quick initial
                        estimation of a mutation’s (de)stabilizing effect, without
                        having to wait for the longer full minimization process.
                        <br><br>
                        RaSP will calculate all mutations for all amino acids in a chain within ~10 minutes. <br><br>
                        If this option is selected, you can press the Ctrl key and left-click on the residue in the
                        viewer.
                        This will open a frame with 20 energy estimates.
                        Be aware that this is an approximation.
                    </p>
                </span></div>
        </li>

    </ul> -->

    <h5>Getting started with <span style="font-variant:small-caps">MutationExplorer</span> </h5>

    <ul class="collapsible">

        <li>
            <div class="collapsible-header grey lighten-4">
                Tutorial 1: Upload structure and mutations: Model upload via PDB ID
            </div>
            <div class="collapsible-body">
                <p>
                    In this tutorial, we will aim to reproduce the aGPCR demo.
                </p>
                <p>
                    In the main web view, click on <a href="{{url_for('submit')}}">upload structure and
                        mutations</a>
                    <br>
                    Structures can be uploaded in three different ways:
                </p>
                <ul class="true-list">
                    <li>directly from the wwPDB via an excisting PDB ID</li>
                    <li>directly from the AlphafoldDB via UniProtKB ID (UniProtKB names are not possible)</li>
                    <li>upload a local PDB:</li>
                    <ul class="true-list" style="margin: 2em; margin-top: 0.5em">
                        <li><span style="font-size: 12px">No HETATM records (can be removed via filter
                                button)</span></li>
                        <li><span style="font-size: 12px">TER after each chain</span></li>
                        <li><span style="font-size: 12px">different and non-empty chain identifiers</span></li>
                        <li><span style="font-size: 12px">no anisotropy entries or multi-state side-chains</span>
                        </li>
                    </ul>
                </ul>

                <p>
                    In this tutorial, we will load a structure from the wwPDB, namely 4DLQ, run a long minimization and
                    calculate the RaSP model. You're selections should look like the following picture:
                    <center>
                        <img src="static/img/tutorial/structure_filled_upload.png" width="60%" />
                    </center>
                </p>

                <p>
                    Clicking the
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="filter structure"
                            style="margin:5px;">
                            filter structure
                        </button>
                    </span>
                    button in the form opens a window where you can remove selected chains from the target PDB and
                    remove hetero atoms, if present.
                    Select the type of minimization to be performed on the provided PDB file. If the structure was
                    not minimized using the ROSETTA force field, download <samp>mut_0.pdb</samp> for a minimized
                    baseline, if the structure was properly minimized, you can go with <em>short</em> minmization.
                    <center>
                        <img src="static/img/tutorial/structure_filter.png" width="80%" />
                    </center>
                </p>

                <p>
                    Enter your E-Mail address to get notified when your minimization is finished.
                </p>
                <p>
                    When you are done, click on
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                            title="Next: input mutations" style="margin:5px;">
                            Next: input mutations
                        </button>
                    </span>
                    <br>
                    The
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="load example"
                            style="margin:5px;">
                            load example
                        </button>
                    </span>
                    on page bottom leads to an example <span style="font-variant:small-caps">MutationExplorer</span>
                    session.
                </p>
                </span>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Tutorial 2a: Upload structure and mutations: define mutations via manual mutation
            </div>
            <div class="collapsible-body">
                <p>
                    There are three ways of providing mutational data:
                </p>
                <ul class="true-list">
                    <li>Manual mutation definition</li>
                    <li>Sequence alignment</li>
                    <li>Target sequence</li>
                </ul>

                <p>
                    In this tutorial, we will focus on a manual mutation definition.
                </p>

                <p>
                    By clicking on the <b>Manual mutation definition</b>, a tab will open that gives information about
                    available PDB chains and amino acids that are within the uploaded PDB, information about the amino
                    acid to one letter code translation, an info button with further explanation and the input form for
                    the mutations.
                    <br>
                    In our case, the
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="btn modal-trigger grey lighten-2"
                            title="available PDB chains" style="margin:5px;">
                            available PDB chains
                        </button>
                    </span>
                    button provides the following information:
                    <center>
                        <img src="static/img/tutorial/tut2_available_chains.png" />
                    </center>
                </p>

                <p>
                    The mutation input field requires a specific format, following a comma separated list of:
                </p>
                <ul class="true-list" id="input-list">
                    <li>
                        A comma separated list of:
                        <span style="font-family: FreeMono;"> CHAIN ':' RESIDUE-ID TARGET-AA-TYPE</span>
                    </li>
                    <li>
                        Example:
                        <span style="font-family: FreeMono;">A:23G,B:123A</span>
                </ul>
                <p>
                    This
                    mutates the residue in chain A with the PDB residue ID 23 to glycine and in chain B the residue with
                    residue ID 123 to alanine. Note that the all mutations will be done in the same model, multiple
                    individual mutations are best done via the explorer interface.
                </p>

                <p>
                    In our example case, we will input <code>A:804S</code>, a mutation in chain A, residue number 804,
                    with the substitution on Serine (S). Multiple mutations follow the same scheme and need to be
                    separated by a comma.
                </p>
                <p>
                    The resulting scheme should look like the picture below:
                    <center>
                        <img src="static/img/tutorial/tut2_define_mutation.png" width="60%" />
                    </center>
                </p>

                <p>
                    Click on
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                            title="submit and launch explorer" style="margin:5px;">
                            submit and launch explorer
                        </button>
                    </span>
                    to proceed to calculation.
                    <br>
                    You will be forwarded to the status page. As the structure is already in the database and
                    has therefore been minimized, the runtime should be around 10-15 minutes. Note that the waiting time
                    can get reduced as the minimization and RaSP calculation already starts after the structures has
                    been uploaded/selected.
                </p>

                <p>
                    When this is finished, you will be re-directed to the <span class="mut-ex">MutationExplorer</span>
                    window. The result is displayed below:
                    <center>
                        <img src="static/img/tutorial/tut2_result.png" width="100%" />
                    </center>
                    For explanation and guidance about usage of the result viewer and interpretation of the results, see
                    the section below.
                </p>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Tutorial 2b: Upload structure and mutations: define mutations via sequence alignment
            </div>
            <div class="collapsible-body">
                <p>
                    If you like to mutate a structure into another structure via a sequence alignment, you can upload
                    such ClustalW formatted alignment. One of the sequences in the alignment has to match <b>exactly one
                        sequence</b> or be a subsequence in the PDB that you uploaded in previous step. Since one
                    alignment is associated with one chain, you can upload up to three alignments. If you need to mutate
                    more than three chains, you can upload target sequences for the other chains or define mutations
                    manually.
                </p>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Tutorial 2c: Upload structure and mutations: define mutations via target sequence
            </div>
            <div class="collapsible-body">
                <p>
                    Provide the sequence for each chain of the uploaded PDB that it will be mutated to.
                    The target chain sequences as a chain of one-letter sequences can be provided as FASTA files (top),
                    pasted directly (center) or directly fetched from UniProtKB (bottom). For each sequence, select the
                    corresponding PDB chain from the dropdown menu. You can also combine different uploads for
                    multi-chain proteins (e.g. Fasta for Chain A, Paste sequence for Chain B). Please ensure that the
                    uploaded/pasted sequences <b>exactly</b>
                    match the sequence in length (check your PDB!) and avoid sequence offsets resulting in mutation of
                    all residues.
                </p>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Tutorial 3: Select structures from UniProt IDs and mutations via a sequence alignment
            </div>
            <div class="collapsible-body">
                <p>
                    <span style="font-size: larger; font-weight: bold;">Model upload via UniProt ID</span>
                    <br><br>

                    For this tutorial, we will select UniProt accession <b>AGRL1_RAT</b> with <b>ID O88917</b>.
                    Therefore, fill in the ID, select a long minimization and check for the RaSP Model. The input form
                    should look like this:
                    <center>
                        <img src="static/img/tutorial/tut4_upload.png" width="60%" />
                    </center>
                    Press the
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                            title="Next: input mutations" style="margin:5px;">
                            Next: input mutations
                        </button>
                    </span>
                    button to proceed to the define mutations tab.
                    <br>
                    In the meantime, an AlphaFold structure will be
                    fetched and a minimization will be performed if the ID is not yet in the database.
                </p>

                <p>
                    <span style="font-size: larger; font-weight: bold;">Define mutations via a target sequence</span>
                    <br><br>
                    Next, we fill select the field <b>Target sequence</b> where we will inset a sequence via the upload
                    or paste option.
                </p>
                <p>
                    In general, the user needs to provide the sequence for each chain of the uploaded PDB to be mutated.
                    The target chain sequences as a chain of one-letter sequences can be provided as FASTA files (top),
                    pasted directly (center) or directly fetched from UniProtKB (bottom). For each sequence, select the
                    corresponding PDB chain from the dropdown menu. You can also combine different uploads for
                    multi-chain proteins (e.g. Fasta for Chain A, Paste sequence for Chain B). Please ensure that the
                    uploaded/pasted sequences <b>exactly</b> match (or is a subsequence of) the sequence in length
                    (check your PDB!) and avoid sequence offsets resulting in mutation of all residues.
                </p>
                <p>
                    In this tutorial, the sequence is a modification of <b>O94910|AGRL1_HUMAN</b>, including:
                </p>
                <ul class="true-list">
                    <li>E701del</li>
                    <li>P1349del</li>
                    <li>G1360del</li>
                    <li>V1146_P1189insAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA</li>
                </ul>
                <p>
                    It is not possible to specify the
                    UniProt ID <b>O94910</b> directly, as we need to ensure that the input sequence has the same length
                    as the sequence in the input structure (<b>O88917|AGRL1_RAT</b>).
                    <!-- Please download the sequence here (O94910-mod.fasta). -->
                </p>

                <p>
                    If the sequence has been pasted in, it will look like the following picture:
                    <center>
                        <img src="static/img/tutorial/tut4_fasta.png" width="80%" />
                    </center>
                </p>
                <p>
                    Next, press
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                            title="submit and launch explorer" style="margin:5px;">
                            submit and launch explorer
                        </button>
                    </span>
                    .
                </p>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Tutorial 4: Perform interface mutations
            </div>
            <div class="collapsible-body">
                <p>
                    In this tutorial, we want to show how to generate interface mutations, e.g. to modify the interface
                    of two different proteins. It will give us the binding score of the 2 proteins. The example we will
                    use is the structure 3SN6, where a beta-adrenergic receptor is interacting with the Gs-protein.
                    Please note, that it is not possible to modify DNA/RNA or substrate interactions.
                </p>
                <p>
                    First, select 3SN6 in your structure upload, remove the hetero atoms via the filter structure option
                    and remove chains B, G and N.
                    <center>
                        <img src="static/img/tutorial/tut5_filter.png" width="80%" />
                    </center>
                    Now select a long Minimization and calculate the RaSP model. Your input form should look like the
                    following:
                    <center>
                        <img src="static/img/tutorial/tut5_upload.png" width="60%" />
                    </center>
                </p>
                <p>
                    Press
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                            title="Next: input mutations" style="margin:5px;">
                            Next: input mutations
                        </button>
                    </span> to proceed.
                    <br>
                    Now we will select the "Interface Score Definition" tab. There you have 3 options:
                </p>
                <ul class="true-list">
                    <li>No calculation</li>
                    <li>Define manual calculation</li>
                    <li>Complete calculation</li>
                </ul>
                <p>
                    We will in this tutorial define the interface manually. If you want to have further explanation
                    about the other modes, check out the FAQ or Documentation.
                    <br>
                    For this, select Chain R and N to be on the Left and Chain A, B and G on the Right.
                    <center>
                        <img src="static/img/tutorial/tut5_interface_score.png" width="80%" />
                    </center>
                </p>
                <p>
                    Next, press
                    <span class="center-align toggle-card">
                        <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                            title="submit and launch explorer" style="margin:5px;">
                            submit and launch explorer
                        </button>
                    </span>
                    <br>
                    This run will take approximately 40 minutes before you are forwarded to the result page. You can
                    speed up the process by not checking RaSP calculations.
                </p>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Tutorial 5: Homology modeling of multiple states of a protein
            </div>
            <div class="collapsible-body">
                <p>
                    What AlphaFold is not able to do is to model different states of a protein. This can be achieved
                    using <span class="mut-ex">MutationExplorer</span>. For many classes and families of proteins,
                    multiple states are available in the PDB. Different states can be modeled By selecting different
                    PDBs as base structure and follow these steps:
                </p>
                <ul class="true-list">
                    <li>
                        Under <b>upload structure and mutations</b> enter the PDB ID of the base structure or upload a
                        model from your machine
                    </li>
                    <li>
                        Ideally select the long minimization (for homology modeling the RaSP option can be disabled) and
                        click the <b>next</b> button
                    </li>
                    <li>
                        For uploading the sequence to which you want to modify the base structure to there are multiple
                        choices:
                        <ul class="true-list">
                            <li>
                                Paste or upload it under <b>target sequence</b> (when the target is rather similar to
                                your base structure / template)
                            </li>
                            <li>
                                If your target sequence has an Uniprot ID you can enter that
                            </li>
                            <li>
                                You will have to indicate to which chain in the base structure the sequence shall be
                                applied (available chains are listed)
                            </li>
                            <li>
                                If it is a very remote relationship, follow the AlignMe tutorial further down
                            </li>
                            <li>
                                for other cases either also use AlignMe, or upload an alignment from another source via
                                <b>sequence alignment</b>
                                <!-- TODO proper link -->
                            </li>
                        </ul>
                    </li>
                </ul>
                <p>
                    The part that cannot be done by <span class="mut-ex">MutationExplorer</span> is the minimization of
                    the model. The user can choose between different possibilities, e.g. using Rosetta as explained
                    above, or by performing MD simulations and many others. Whatever approach the user choosed, it
                    requires sufficient sampling and thus has to be done locally.
                </p>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Tutorial 6: Upload VCF (sequencing data)
            </div>
            <div class="collapsible-body">
                <p>
                    vCF, the tab-delimited Variant Call Format, is incredibly powerful and flexible. As example of the
                    minimal input format used to upload SNPs to <span class="mut-ex">MutationExplorer</span>, see the
                    sample file
                    <a
                        href="https://proteinformatics.uni-leipzig.de/mutation_explorer/downloads/rtel1/rtel1.vcf">RTEL1.vcf</a>,
                    described in our manuscript and the documentation (click on the name to download it).
                </p>
                <p>
                    This example file contains the following input:
                    <center>
                        <img src="static/img/tutorial/vcf.png" />
                    </center>
                </p>
                <p>
                    This tutorial lists the steps you will take
                    to upload RTEL1.vcf and replicate the example page
                    <a href="https://proteinformatics.uni-leipzig.de/mutation_explorer/examples/rtel1/">RTEL1 <span
                            class="mut-ex">MutationExplorer</span> results</span></a>
                    results. Under the hood, the sequence data will be translated into mutations, an AlphaFold structure
                    will be generated of the associated protein and the variants will be highlighted. Further (if
                    selected), the RaSP model will calculate energy differences for future variants.
                </p>

                <p>
                    <u>RTEL1 Tutorial Steps</u>
                </p>

                <ol>
                    <li> From the home page, click on
                        <span class=" center-align toggle-card">
                            <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                                title="download all" style="margin:0px;">
                                upload vcf
                            </button>
                        </span>
                        and you will see a new form with a detailed explanation of <span
                            class="mut-ex">MutationExplorer</span>’s VCF workflow and limitations.
                        You can select different minimization length as described in the documentation, calculate
                        energies via the RaSP model or calculate interface scores. Optionally, an email can be added to
                        obtain notifications when a minimization job has been completed.

                        <center>
                            <img src="static/img/tutorial/vcf_upload_form.png" width="80%" />
                        </center>
                        <br>
                    </li>

                    <li style="margin-bottom: 20px;">
                        In this tutorial, we select <em>long minimization</em> to replicate the energy values of our
                        example page. If <em>quick</em> has been selected, be aware that numerical precision might
                        differ. <em>None</em> is fastest, but energetic calculations will be meaningless.
                        <br>
                        Click the
                        <span class="grey-text text-darken-1">
                            <button class="waves-effect waves-light btn modal-trigger grey lighten-2">i</button>
                        </span>
                        at right for more details.
                    </li>

                    <li style="margin-bottom: 20px;">
                        Click the <img src="static/img/hook.png"> under <em>RaSP</em> f you wish this stability
                        prediction calculation.<br>
                        Click the
                        <span class="grey-text text-darken-1">
                            <button class="waves-effect waves-light btn modal-trigger grey lighten-2">i</button>
                        </span>
                        at right for more details.
                    </li>

                    <li style="margin-bottom: 10px;">
                        Now click on the
                        <span class="center-align toggle-card">
                            <button class="waves-effect waves-light btn modal-trigger" title="download all"
                                style="margin:0px;">
                                Upload a VCF file
                            </button>
                        </span>
                        button and include the above mentioned sample VCF file. After this step, the page should look
                        like this:
                        <center>
                            <img src="static/img/tutorial/vcf_pre_submit.png" width="60%" />
                        </center>
                    </li>

                    <li>
                        Click
                        <span class="center-align toggle-card">
                            <button class="waves-effect waves-light btn modal-trigger" title="download all"
                                style="margin:0px;">
                                submit
                            </button>
                        </span>
                        to start the calculations. The webpage tab at the top of the browser window will immediately
                        indicate a refresh in progress. Very soon you are transported to a new page showing <em>Status
                            of Job nnnnn</em>. Be patient as the calculation proceeds. It is a good idea to right-click
                        on <em>Result Page</em> for later references. However, this URL will not be active until the
                        various calculations complete (indicated by the message <em>"This url does not belong to the
                            app."</em>). At that time, you will be automatically redirected to this same <em>Result
                            Page</em>. Your patience will be well rewarded.
                        <br><br>
                        See here an example of the status file:
                        <center>
                            <img src="static/img/tutorial/status_job.png" width="100%" />
                        </center>
                        You will obtain further information on the calculation status, e.g.:
                        <ul class="true-list">
                            <li>Start Calculation</li>
                            <li>Start fixbb for mut_0_1</li>
                            <li>fixbb for mut_0_1 done</li>
                            <li>Start RaSP calculation for mut_0_1 with chain A</li>
                        </ul>
                        The overall runtime should take approximately 10-15 minutes.
                    </li>
                </ol>

                <p>
                    <u>Troubleshooting your VCF uploads</u>
                    <br>

                    The processes between the submission of your VCF file, and the display of an AlphaFold model are
                    complex. Your VCF upload will be processed by the Ensembl Variant Effect Predictor (VEP), running
                    locally on the <span class="mut-ex">MutationExplorer</span> server. If your results are unexpected,
                    please follow our guidelines in the documentation or FAQ.
                </p>
                <p>
                    After 10 minute waiting time, you will be forwarded to the result page:
                    <center>
                        <img src="static/img/tutorial/vcf_result.png" width="100%" />
                    </center>
                    For guidance and information about the setup of the result visualization and an interpretation of
                    the results, please see the section below.
                </p>

                <p>
                    While seeing missense variants from the public VEP website is encouraging, other difficulties may
                    remain.
                    <br>
                    For example:
                <ul class="true-list" id="input-list">
                    <li>
                        <span style="font-variant:small-caps">MutationExplorer</span> can only analyze one protein at a
                        time. You will need to split your VCF file and make separate runs, if your VCF file generates
                        missense variations for multiple genes.
                    </li>
                    <li>
                        The VEP often returns Ensembl transcript Ids which do not map to Swiss-Prot curated, canonical
                        UniProt transcript IDs. Without a UniProt ID, <span class="mut-ex">MutationExplorer</span>
                        cannot retrieve an AlphaFold model.
                        You may manually check cross-references between ENST… transcript IDs and Uniprot IDs at <a
                            href="https://ensembl.org">ensembl.org</a> and <a
                            href="https://uniprot.org">uniprot.org</a>, respectively.
                    </li>
                    <li>
                        Inputting GRCh37/hg19 genomic positions, which are indistinguishably formatted, will result in
                        wildly incorrect results. You must first “liftover” such coordinates, by converting your VCF
                        coordinates to BED format, then submitting them to <a
                            href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">https://genome.ucsc.edu/cgi-bin/hgLiftOver</a>.
                        As you use these kinds of tools, remember that <span class="mut-ex">MutationExplorer</span>
                        requires GRCh38.
                    </li>
                </ul>

            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-4">
                Tutorial 7: Start from the AlignMe webserver - discover sequence alignments in 3D
            </div>
            <div class="collapsible-body"><span>
                    <a href="https://www.bioinfo.mpg.de/AlignMe/AlignMePW.html">AlignMe</a> is a software package and
                    webserver
                    for detecting similarities between proteins, which are too subtle to be detected on the sequence
                    level using
                    standard methods.
                    <br><br>
                    <a
                        href="https://www.bioinfo.mpg.de/AlignMe/AlignMePW.html">https://www.bioinfo.mpg.de/AlignMe/AlignMePW.html</a>
                    <br><br>
                    Follow these steps:
                    <ul class="true-list" id="input-list">
                        <li> paste two sequences in Fasta format into the windows </li>
                        <li> if the sequences are rather similar select the 'fast' alignment</li>
                        <li> if there is little similarity on a sequence level, use one of the three other modes.</li>
                        <li> submit</li>
                        <li> on the result page, click the button to <span
                                style="font-variant:small-caps">MutationExplorer</span>
                        </li>
                    </ul>
                    <br>
                    Now you are on a form of <span style="font-variant:small-caps">MutationExplorer</span> where you can
                    upload
                    one or two structures on which the alignments are mapped.
                    <ul class="true-list" id="input-list">
                        <li> define base and target sequence from your alignment: the dropdown will display all fasta
                            headers that
                            you entered on the AlignMe website </li>
                        <li> upload structure for the base sequence (required) </li>
                        <li> define chain in PDB (crucial when you have e.g. homomultimers, otherwise the chain is
                            determined
                            automatically)</li>
                        <li> optional: upload structure for target sequence </li>
                        <li> define chain in PDB (crucial for homomultimers)</li>
                        <li> select energy minimization (default is 'none') </li>
                        <li> select RaSP (default is 'off') </li>
                        <li> submit</li>
                    </ul>
                    If you select an energy minimization or RaSP calculation, you will have to wait accordingly.<br><br>
                    The result window will display different things according to what you uploaded.
                    <br>
                    <u>Single PDB</u>
                    <ul class="true-list" id="input-list">
                        <li> base PDB is mutated to target sequence </li>
                        <li> structure is colored by sequence conservation </li>
                        <li> coloring by energy is available (quality depends on selection of minimization)</li>
                        <li> coloring by hydrophobicity is available </li>
                    </ul>
                    <br>
                    <u>Two PDBs</u>
                    <ul class="true-list" id="input-list">
                        <li> PDBs are structurally aligned (superimposed) according to sequence alignment </li>
                        <li> structure is colored by sequence conservation </li>
                        <li> coloring by energy is available (quality depends on selection of minimization)</li>
                        <li> coloring by hydrophobicity is available </li>
                        <li> you can move the structures independently by ... MICHELLE </li>
                    </ul>
                </span></div>
        </li>

        <!-- <li>
            <div class="collapsible-header grey lighten-5">

            </div>
            <div class="collapsible-body">
                <p>

                </p>
            </div>
        </li> -->

    </ul>

    <br>

    <h5>The main stage: <span class="mut-ex">MutationExplorer</span> result window</h5>

    <ul class="collapsible">

        <li>
            <div class="collapsible-header grey lighten-4">
                General
            </div>
            <div class="collapsible-body">
                <p>
                    In the explorer view, you will be presented with information about your set of mutations, relative
                    per-residue energies and visualization of the protein.
                    We will first have a closer look to the info bar on the left. It is divided into 3 sections, Select,
                    Mutate and Info:
                    <center>
                        <img src="static/img/tutorial/tut2_result.png" width="100%" />
                    </center>
                </p>
            </div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-4">
                Left side (info bar)
            </div>
            <div class="collapsible-body">
                <p>
                    We will first have a closer look to the info bar on the left. It is divided into 3 sections, Select,
                    Mutate and Info:
                    <center>
                        <img src="static/img/tutorial/tut2_result_left.png" width="50%" />
                    </center>
                </p>

                <p>
                    <span style="font-weight: bold; font-size: larger;">Select</span>
                </p>
                <ul class="true-list">
                    <li style="margin: 10px;">
                        <b>Tree View:</b> Outlines the mutations that were performed, where <em>mut_0</em> will be the
                        unmutated, minimized PDB file that serves as the basis of mutations. You can select entries in
                        the tree view for individual mutations. Every newly generated PDB will be a new entry. If you
                        performed your own calculation, you will see only 2 entries, namely <em>mut_0</em> and
                        <em>mut_0_1</em>. If you selected the session from the examples, you will see 3 further entries,
                        namely <em>mut_0_2-4</em>. In both cased, the current visualized entry (<em>mut_0_1</em>) is
                        marked in bold. Go ahead and try switching between the different selections, the orientation and
                        zoom of the protein will remain the same.
                    </li>
                    <li style="margin: 10px;">
                        <b>Color by Dropdown menu:</b> provides different coloration methods:
                        <ul class="true-list">
                            <li>Absolute energy</li>
                            <li>Energy difference to parent</li>
                            <li>Absolute hydrophobicity</li>
                            <li>Difference in hydrophobicity to parent</li>
                            <li>Conservation</li>
                            <li>Absolute interface score</li>
                            <li>Difference in interface score to parent</li>
                        </ul>
                        for the coloration of the protein in the main window. You can find further information about
                        them in the documentation and FAQ. For this tutorial, we will select energy difference to parent
                        (default selection).
                    </li>
                    <li style="margin: 10px;">
                        <b>Currently displayed</b> shows the current selection fo the entry and the coloring scheme.
                        <ul class="true-list">
                            <li>absE - Absolute energy</li>
                            <li>diffE - Energy difference to parent</li>
                            <li>HyPh - Absolute hydrophobicity</li>
                            <li>diffHyPh - Hydrophobicity difference to parent</li>
                            <li>cons - Sequence conservation</li>
                            <li>diffIF - Interface score difference to parent</li>
                            <li>IF - Absolute interface score</li>
                        </ul>
                    </li>
                    <li style="margin: 10px;">
                        <span class="center-align toggle-card">
                            <button data-target="modal3" class="waves-effect waves-light btn modal-trigger"
                                title="download all" style="margin:0px;">
                                download all
                            </button>
                        </span>
                        pack all PDB files into a ZIP archive for downloading
                    </li>
                </ul>

                <p>
                    <span style="font-weight: bold; font-size: larger;">Mutate</span>
                </p>
                <ul class="true-list">
                    <li style="margin: 10px;">
                        <b>Enter mutations field</b>: allows for further mutations to be entered, with the root being
                        the currently active protein (printed in bold in the tree view). The <b>info</b> field below
                        provides info on the currently selected protein.
                    </li>
                    <li style="margin: 10px;">
                        <b>Get notified after job completion</b>: you input your email to get a notification when the
                        run is completed.
                    </li>
                </ul>
                <p>
                    If you are using the example, you can skip this part, but feel free to try out different mutations
                    yourself. We will now perform 3 rounds of mutations. We want them to be introduced to our original
                    (minimized) pdb, so we first select <em>mut_0</em> in the selection tree. Next, we enter the first
                    additional mutation into the mutation field: <em>A:815S</em>. The left panel should now look like
                    this:
                    <center>
                        <img src="static/img/tutorial/tut_vis_mutate1.png" width="50%" />
                    </center>
                    Press the submit button to initiate the mutations.
                    <br>

                    After this, we will do the same for mutation <em>A:806A</em> and <em>A:803Y</em>. Do not forget to
                    click on the initial structure <em>mut_0</em> before submitting the run.
                    <br><br>
                    Finally, we will end up with the same setting as the example aGPCR.
                </p>
                <p>
                    Now we will do a final set of mutations: we will select our first variant <em>mut_0_1</em> and input
                    the 3 mutations <em>(A:815S,A:806A,A:803Y)</em>. This will introduce all 4 mutations into the
                    wild-type.
                    <br><br>
                    Your panel should look like the following:
                    <center>
                        <img src="static/img/tutorial/tut_vis_multi_mutation.png" width="50%" />
                    </center>
                </p>

                <p>
                    <span style="font-weight: bold; font-size: larger;">Info</span>
                    <br>
                    <br>
                    Finally, at the bottom we see the info panel. This table is showing information about your current
                    selection: Total Energy (Rosetta Energy units),total energy difference to parent, the parent
                    structure, introduced mutations, the original file name and the exploration ID. For this tutorial,
                    it might look similar to this:
                    <center>
                        <img src="static/img/tutorial/tut_vis_info.png" width="50%" />
                    </center>
                    Please note that the exploration ID and the energies might be different due to different rotamers
                    selected during minimization.
                </p>
            </div>
        </li>


        <li>
            <div class="collapsible-header grey lighten-5">
                Right side (main visualization window)
            </div>
            <div class="collapsible-body">
                <p>
                    At the top center (main part), we find the Mol* visualization of the protein with a blue-white-red
                    color gradient, where blue represents negative values and red positive values in the default color
                    setting. The mutated residues are highlighted with a red translucent sphere. Hovering over the
                    cartoon will highlight residues and display a tooltip with further information in the bottom right
                    corner.
                    <br><br>
                    You can see one example here:
                    <center>
                        <img src="static/img/tutorial/tut_vis_tooltip.png" width="100%" />
                    </center>
                </p>
                <p>
                    At the bottom, you can find the MDSrv Window: By default, it shows a sequence alignment of your root
                    and mutated protein with the currently mutated residue highlighted in red. Hovering over cartoon
                    residues will also highlight the residue in the alignment and vice-versa. It will further display a
                    tool-tip on the right side with information about the highlighted residue.
                    <br><br>
                    You can see one example here:
                    <center>
                        <img src="static/img/tutorial/tut_vis_seq.png" width="100%" />
                    </center>
                </p>
                <p>
                    Clicking on any highlighted residue will zoom in, showing the selected residue and its surroundings
                    in a ball-and-stick view. Different carbon colors indicate protein chains, otherwise atoms are
                    colors by type. Dashed lines indicate residue-residue interactions like hydrogen bonds.
                    <br><br>
                    Try it out - you might see something similar to this:
                    <center>
                        <img src="static/img/tutorial/tut_vis_zoom.png" width="100%" />
                    </center>
                    To get out of this zoom, click on some of the background. This will bring you back to the original
                    view.
                </p>
                <p>
                    <b>Change visualisation</b>
                    <br>
                    Now we want to modify the visualization. Click therefore on the bottom-right button in the alignment
                    section. The result will look similar to this:
                    <center>
                        <img src="static/img/tutorial/tut_vis_change_vis.png" width="100%" />
                    </center>
                    Here you can do adjustments to the visualization, as changing e.g. the sphere size or color of each
                    individual mutation. If you want to reset those adjustments, simply reload the page.
                    If you want to learn more about how to adjust the visualization, please go the the Documentation.
                </p>
            </div>
        </li>

        <!-- <li>
            <div class="collapsible-header grey lighten-4">
                Adjusting visualization
            </div>
            <div class="collapsible-body"><span>
                    <p>
                    <ul class="true-list" id="input-list"><br>
                        You can adjust the metric that serves as the basis for coloration in the <i>"Color by:"</i>
                        Dropdown menu.
                        By default, the colormap will be from -2 to 2 in a blue-white-red gradient
                        <li style="margin: 10px;">Click on the <span style="font-family:Arial"><b>Change
                                    Visualization</b></span>
                            tab in the bottom right of the main window to adjust the color scale range and the mutation
                            markers</li>
                        <li style="margin: 10px;">First select for which structure you want to change the range of the
                            color scale
                            by choosing the corresponding file in the Structure drop-down menu. On the left and right
                            side of the
                            color scale, change the numbers and confirm with enter. The color scale will be adjusted
                            upon clicking
                            <span style="font-family:Arial"><b>Apply</b></span> below the scale
                        </li>
                        <li style="margin: 10px;">The mutation marker color and size can be adjusted via the slider and
                            color
                            window. The marker can be turned off by clicking on <span style="font-family:Arial"><b>✓
                                    On</b></span> in
                            the Show Field, setting it to <span style="font-family:Arial"><b>X Off</b></span></li>
                        <li style="margin: 10px">It is possible to hide all mutation spheres at once. Click the <span
                                style="font-family:Arial"><b>Disable all mutation spheres</b></span> button in the
                            Mutation Markers
                            panel. To show all mutation spheres again, click the <span
                                style="font-family:Arial"><b>Enable all
                                    mutation spheres</b></span> button</li>
                        <li style="margin: 10px">If you are importing multiple structures into a scene, it may be
                            beneficial to hide
                            a specific structure. Select the structure you want to hide from the Structure drop-down
                            menu. Then click
                            the <span style="font-family:Arial"><b>Hide Structure</b></span> button at the bottom. To
                            make the
                            structure visible again, simply click the <span style="font-family:Arial"><b>Show
                                    Structure</b></span>
                            button</li>
                    </ul>
                    </p>
                </span></div>
        </li> -->

        <li>
            <div class="collapsible-header grey lighten-5">
                Visualize RaSP mutations
            </div>
            <div class="collapsible-body">
                <p>
                    To visualize the effect for all potential other mutations, activate the RaSP panel. Click therefore
                    at the red mouse at the top-left of the visualization panel:
                    <center>
                        <img src="static/img/tutorial/tut_vis_red_mouse.png" width="20%" />
                    </center>
                    After clicking, this will become green. Now you can click on a residue within the sequence alignment
                    and a new window will open with the scores for the single substitutions at this position. A low or
                    negative score means stabilization, a high score means that his substitution will be destabilizing.
                    <center>
                        <img src="static/img/tutorial/tut_vis_rasp.png" width="100%" />
                    </center>
                    When clicking on Serin 815, we see that a Tryptophan (W) mutation gives the most stabilizing effect.
                    Let's see if our structure gets more stabilized with this <em>A:815W</em> mutation (paste it into
                    the form and press submit). If you do not know how to initiate mutations in the viewer, please see
                    above or check out the documentation.
                </p>
                <p>
                    As a result, we can see that the energy difference to the WT got reduced in comparison to
                    <em>mut_0_1_1</em>. When switching now to coloring according to the absolute energy, we can see the
                    red parts, that could be targets for further modification towards a stabilized structure:
                    <center>
                        <img src="static/img/tutorial/tut_vis_absE.png" width="100%" />
                    </center>
                </p>
            </div>
        </li>
    </ul>

    <!-- <li>
            <div class="collapsible-header grey lighten-5">
                Backmutating PDB structures to wildtype sequence for MD simulations
            </div>
            <div class="collapsible-body"><span>
                    <p>
                        Structures deposited in the PDB are very often modified in order to solve them experimentally.
                        Using <span style="font-variant:small-caps">MutationExplorer</span>, it is a simple two-step
                        process to
                        mutate them back to their native sequence.
                        Thus it is common step for simulating proteins to back-mutate PDB.
                    </p>
                    <p>
                        However, it should be noted that at its current state <span
                            style="font-variant:small-caps">MutationExplorer</span> is not able to complete the
                        structure by modeling
                        missing parts in the PDB.
                    </p>
                </span></div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-4">
                Multi-state modeling of homologs via Molecular Dynamics simulation ensembles
            </div>
            <div class="collapsible-body"><span>
                    Molecular dynamics simulation can generate ensembles of functionally relevant states. <span
                        style="font-variant:small-caps">MutationExplorer</span> maps those ensembles to close or distant
                    homologs.
                    For close homologs, an alignment will be automatically generated internally. Alternatively, a
                    pre-calculated
                    sequence alignment can be provided. Where more refined alignments are required, the server accepts
                    results
                    forwarded from the AlignMe website (<a
                        href="https://doi.org/10.1093/nar/gkac391">https://doi.org/10.1093/nar/gkac391</a>). Such
                    sequence
                    alignments carry nuanced details and <span style="font-variant:small-caps">MutationExplorer</span>
                    drills down
                    to reveal them all.
                </span></div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-5">
                Aiding design for experimental studies for protein characterization
            </div>
            <div class="collapsible-body"><span>
                    When designing proteins, the effect of the variant is often only known after applying of an external
                    software,
                    via multiple design rounds. After each of those, the structures can be inspected. The power of the
                    <span style="font-variant:small-caps">MutationExplorer</span> lies in the possibility to do this
                    directly in the
                    viewer and select novels rounds of mutations immediatley and due to their influence on the
                    structure.
                </span></div>
        </li>

        <li>
            <div class="collapsible-header grey lighten-4">
                (De-) Stabilize a protein-protein interface
            </div>
            <div class="collapsible-body"><span>
                    3D mapping of energies is not limited to monomers. Over single or multiple rounds of mutations,
                    protein
                    complexes or protein-protein interfaces can flexibly be designed to achieve (de)stabilization. The
                    effect of
                    each design choice can be explored visually.
                </span></div>
        </li> -->

    <br>

    <h5>Happy mutating!</h5>

    <br>
    <br>
    <br>
    <br>
    <br>
    <p>
        If you are interested in further modifications, or result interprestion, please check out the manual, FAQ or
        reach out to us.
    </p>

    <p> For further information, see the <a href="faq">FAQs</a>.</p>
    <p> Also check out the <a href="examples">example section</a>.</p>

    <br>
    <br><br>
</center>

{% endblock %}