{% extends "base.html" %}

{% block content %}

{% block head %}
<style>
  #input-list li {
  margin-left: 1em;
  padding-left: 1em;
  list-style-type: disc !important;
  line-height:140%
    }

</style>
{% endblock %}


<center>
<h1>Tutorial for <span  style="font-variant:small-caps">MutationExplorer</span> </h1>
<div><img src="static/img/workflow.png" width="100%;"></div>
</center>

<!--

<p>
  The server lets users mutate proteins in
  various ways and study mutation effects in 3D.
  The only required inputs are a structural model and a list of mutations of
  interest.
  Alternatively, the user can upload human SNPs in the VCF format.
  <br>
  You can click on the       <span class="center-align toggle-card"> <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="Further information" style="margin:5px;">i</button> </span> buttons at each field in the forms to get further information.
  <div id="modal3" class="modal">
    <div class="modal-content">
      This button has no function in the tutorial.
      <b>Happy mutating!</b>
      <p> </p>
    </div>
    <div class="modal-footer">
	    <a href="#!" class="modal-close waves-effect waves-green btn-flat">Close</a>
    </div>
  </div>
</p>
-->
  <br><br>

<h3>Overview</h3>
<p> There are three ways of using <span  style="font-variant:small-caps">MutationExplorer</span>.<br>
  First, you can create sequence alignments using the AlignMe server and forward the results to <span  style="font-variant:small-caps">MutationExplorer</span>:<br>
  1) <a href="tutorial#alignme">AlignMe</a> <br><br>
  Then there are
  two main modes of usage that are related to the tabs in the top menu.<br><br>
  First, one can upload sequencing data: <br>
  2) <a href="tutorial#vcf">"Upload VCF"</a> <br><br>
  
  Second, you can "upload structure and mutations" as explained in:<br>
  3) <a href="tutorial#upload">Upload model</a> <br>
  4) <a href="tutorial#mutate">Define mutations</a> <br>
  Both lead to the explore window: <br>
  5) <a href="tutorial#explore">MutationExplorer</a> <br>
  General remarks:<br>
  6) <a href="#mini">Energy minimization and database</a> <br>
  7) <a href="tutorial#specific">Specific mutations</a> <br>
</p>

<div id="alignme" name="alignme" style="top-margin:2cm;"></div>
<br>
<br>
<h4>1) AlignMe </h4>

<a href="https://www.bioinfo.mpg.de/AlignMe/AlignMePW.html">AlignMe</a> is a software package and webserver for detecting similarities between proteins, which are too low to be detected on the sequence level using standard methods.
Follow these steps:
<ul id="input-list">
  <li> paste two sequences in Fasta format into the windows under 1.</li>
  <li> if the sequences are rather simple select the 'fast' alignment under 2.</li>
  <li> if there is little similarity on a sequence level, use one of the three other modes.</li>
  <li> submit</li>
  <li> on the result page, click the button to <span  style="font-variant:small-caps">MutationExplorer</span> </li>
  <li> now a form opens where you can upload one or two structures on which the alignments are mapped. </li>
</ul>


<div id="vcf" name="vcf"></div>
<br><br>
<h4> 2) Upload VCF</h4>
<p> This form allows to upload human SNPs in VCF format. <br>
  The upload and handling is simple, select a file and submit. </p>
  


<ul id="input-list" style="font-size: 14px;border: 1px solid darkgrey ">
  <b>Expected time requirements</b><br>
  
  <li> Initial basic minimization: 1 min </li>
  <li> Mutation and basic minimization: 2 min  </li>
  <li> VCF processing:  5 min  </li>
  Actual times depend on the size of the system and may vary significantly.
</ul>


<div id="upload" name="upload"></div>
<br><br>
<h4>  3) Upload model</h4>
<p>
  In the main web view, click on "upload structure and mutations" Structures can be uploaded in three different ways:
  <ul id="input-list">
    <li>directly from the PDB via PDB ID</li>
    <li>directly from the AlphafoldDB via UniProtKB ID</li>
    <li>upload a local PDB:</li>
    <ul style="margin: 2em; margin-top: 0.5em">
      <li><span style="font-size: 12px">No HETATM records (can be removed via filter button)</span></li>
      <li><span style="font-size: 12px">TER after each chain</span></li>
      <li><span style="font-size: 12px">different and non-empty chain identifiers</span></li>
      <li><span style="font-size: 12px">no anisotropy entries or multi-state side-chains</span></li>    
    </ul>
  </ul>
  
  Clicking the <span class="center-align toggle-card"> <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="filter structure" style="margin:5px;">filter structure</button> </span> button in the form opens a window where you can remove selected chains from the target PDB and remove hetero atoms, if present.
Select the type of minimization to be performed on the provided PDB file. If the structure was not minimized using the ROSETTA force field, download mut_0.pdb for a minimized baseline, if the structure was properly minimized, you can go with "short" minmization.
<br><br>
Enter your E-Mail address to get notified when your minimization is finished.<br><br>
When you are done, click on <span class="center-align toggle-card"> <button data-target="modal3" class="waves-effect waves-light btn modal-trigger" title="Next: input mutations" style="margin:5px;">Next: input mutations</button> </span><br>
The <span class="center-align toggle-card"> <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="load example" style="margin:5px;">load example</button> </span>on page bottom leads to an example <span style="font-variant:small-caps">MutationExplorer</span> session.
  </p>
  
<div id="mutate" name="mutate"></div>
<br><br>
<h4> 4) Define mutations</h4>

<p>
There are three ways of providing mutational data:
<ul id="input-list">
  <b>4.1) Manual mutation definition</b><br><br>
  <span class="center-align toggle-card"> <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="available PDB chains" style="margin:5px;">available PDB chains</button> </span>provides an overview over the available PDB chains in the uploaded PDB,<br><br>
  <span class="center-align toggle-card"> <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="Further information" style="margin:5px;">info</button> </span>explains the mutation syntax,
  <br><br>
  <span class="center-align toggle-card"> <button data-target="modal3" class="btn modal-trigger grey lighten-2" title="one letter code" style="margin:5px;">one letter code</button> </span>provides amino acid one-letter codes.
  <br><br>
  Define a list of point amino acid substitutions with the following syntax:
  <br><br>
  <ul id="input-list">
  <li>A comma separated list of: <span style="font-family: FreeMono;"> CHAIN ':' RESIDUE-ID TARGET-AA-TYPE</span></li>
  <li>Example: <span style="font-family: FreeMono;">A:23G,B:123A</span><br>
  <span style="font-size: 12px; color: #666;">This mutates the residue in chain A with the PDB residue ID 23 to glycine and in chain B the residue with residue ID 123 to alanine.
  </ul><br>
  Note that the all mutations will be done in the same model, multiple individual mutations are best done via the explorer interface.<br><br>

  <b>4.2) Sequence Alignment</b><br><br>

  You can mutate the model that you uploaded in previous step using a sequence alignment in ClustalW format. One of the sequences in the alignment has to match <b>exactly one sequence</b> in the PDB that you uploaded in previous step. Since one alignment is associated with one chain, you can upload up to three alignments. If you need to mutate more than three chains, you can upload target sequences for the other chains or define mutations manually.<br><br>

  <b>4.3) Target Sequence</b><br><br>

  Provide the sequence for each chain of the uploaded PDB that it will be mutated to. The target chain sequences as a chain of one-letter sequences can be provided as FASTA files (top), pasted directly (center) or directly fetched from UniProtKB (bottom)
  <br>
  <b>This is a stub and needs separate testing</b><br>
</ul>
  Click on<span class="center-align toggle-card"> <button data-target="modal3" class="waves-effect waves-light btn modal-trigger" title="submit and launch explorer" style="margin:5px;">submit and launch explorer</button> </span> to proceed to caluclation. When this is finished, you will be re-directed to the <span style="font-variant:small-caps">MutationExplorer</span> window.
</p>



<div id="explore" name="explore"></div>
<br><br>
<h4> 5) <span style="font-variant:small-caps">MutationExplorer</span> Window</h4>
<p>
<ul>
In the explorer view, you will be presented with information about your set of mutations, relative per-residue energies and visualization of the protein.<br><br>
<b>Left side (info bar):</b><br><br>
<ol type="a" >
  <b>Select</b><br>
  <li style="margin: 10px;"> Tree View: Outlines the mutations that were performed, where <span style="font-family: FreeMono; color:darkblue;"><b>mut_0</b></span> will be the unmutated, minimized PDB file that serves as the basis of mutations. You can select entries in the tree view for individual mutations.</li>
  <li style="margin: 10px;">"Color by" Dropdown menu: provides different coloration methods (Absolute Energy, Energy Difference to parent, absolute hydrophobicity, Difference in hydrophobicity) for the coloration of the protein in the main window</li>
  <li style="margin: 10px;"><span class="center-align toggle-card"> <button data-target="modal3" class="waves-effect waves-light btn modal-trigger" title="download all" style="margin:0px;">download all</button> </span> button: pack all PDB files into a ZIP archive for downloading</li>
  <b>Mutate</b><br>
  <li style="margin: 10px;">"Enter mutations" field: allows for further mutations to be entered, with the root being the currently active protein (printed in bold in the tree view). The "info" field onm the bottom provides info on the currently selected protein.</li>
  <b>Info</b><br>
  <li style="margin: 10px;">Table Showing information about your current selection: Total Energy (Rosetta Energy units), the parent structure, introduced mutations and the exploration ID.</li>
</ol><br>
<b>Right side (main window):</b>
<ol type="a">
  <li style="margin: 10px;">A Mol* visualization of the protein with a <i><span style="color:blue">blue</span>-white-<span style="color:red">red</span></i> color gradient, where blue represents negative values and red positive values in the default color setting. The mutated residues are highlighted with a red translucent sphere. Hovering over the cartoon will highlight residues and display a tooltip with further information in the bottom right corner.</li>
  <li style="margin: 10px;">MDSrv Window: By default, shows a sequence alignment of your root and mutated protein with the currently mutated residue highlighted in red. Hovering over cartoon residues will also highlight the residue in the alignment and vice-versa. It will further display a tool-tip on the right side with information about the highlighted residue.</li>
</ol><br>
Clicking on any highlighted residue will zoom in, showing the selected residue and its surroundings in a ball-and-stick view. Different carbon colors indicate protein chains, otherwise atoms are colors by type. Dashed lines indicate residue-residue interactions like hydrogen bonds.<br><br>


  <b>5.1) Adding mutations</b><br>
  <ul id="input-list">
    <li style="margin: 10px;">In the tree view, click on the protein to serve as the root for your next mutation, e.g.  <span style="font-family: FreeMono; color:darkblue;"><b>mut_0_1</b></span>, which is the first mutation entered in the webpage</li>
    <li style="margin: 10px;">In the <b>Mutate</b> field, enter the mutations to be added to the protein. Click on the "info" box for more informations</li>
    <li style="margin: 10px;">Click on <span class="center-align toggle-card"> <button data-target="modal3" class="waves-effect waves-light btn modal-trigger" title="submit" style="margin:5px;">submit</button> </span> and wait for your new protein to be added. The waiting message shows the name of the new mutation. In the tree, the new mutated protein will appear indented under the selected root</li>
  </ul>
  <b>5.2) Adjusting visualization</b><br>
  <ul id="input-list"><br>
  You can adjust the metric that serves as the basis for coloration in the <i>"Color by:"</i> Dropdown menu. By default, the colormap will be from -15 to 15 in a blue-white-red gradient
  <li style="margin: 10px;">Click on the <span style="font-family:Arial"><b>Change Visualization</b></span> tab in the bottom right of the main window to adjust the color scale and the mutation marker</li>
  <li style="margin: 10px;">On the left and right side of the color scale, change the numbers and confirm with enter. The color scale will be adjusted upon clicking <span style="font-family:Arial"><b>Apply</b></span> below the scale</li>
  <li style="margin: 10px;">The mutation marker color and size can be adjusted via the slider and color window. The marker can be turned off by clicking on <span style="font-family:Arial"><b>✓ On</b></span> in the Show Field, setting it to <span style="font-family:Arial"><b>X Off</b></span></li>
</ul>
</p>



<div id="mini" name="mini"></div>
<br><br>
<h4> 6) Energy minimization</h4>
<p>
  The minimization is crucial for the quality of the outcome. The better the structure is energetically minimized, the more reliable the results will be. <br>
  We continuously pre-minimize structures from the PDB using the commandline described below.
  If you specify a PDB ID that is contained in our database, the pre-minimized structure is used. <br>
  Otherwise we offer two modes of side-chain minimization, a short and a long one. The short one should take a few minutes, depending on the size of the protein. The long one may take some hours for large proteins. 
</p>
<h5> AlphaFold and your own models</h4>
<p>These are not pre-minimized and you should follow the steps below. The AlphaFold models should be minimized, but the minimization has to be performed with the same energy functions than we use in the <span class="mut-ex">MutationExplorer</span>, namely Rosetta. </p>

<h5> Do it yourself!</h4>
<p> The minimizations we offer as 'long' and 'short' are both only side-chain optimizations.
  However, we recommend you to perform the following minimization. It is optimizing also the backbone in a limited way (due to '-relax:constrain_relax_to_start_coords'). A free backbone minimization might be leading to problems for e.g. membrane proteins. </p>
<div class="codebox">
  <code>
    PATH_TO_ROSETTA/relax.static.linuxgccrelease   -relax:fast true -relax:cartesian true -score:weights ref2015_cart -use_input_sc -optimization::default_max_cycles 200 -linmem_ig 10 -relax:constrain_relax_to_start_coords -ex1 -ex2  -nstruct 20 -in:file:s  INPUT.pdb -out:pdb -out:prefix  OUTDIR/
  </code>
</div>

<p>
  This commandline will create 20 models, from which you should select the one with the lowest energy. Each PDB contains a line starting with 'pose'. In this line you want the last value, which is the total energy.</p>



<div id="specific" name="specific"></div>
<br><br>
<h4> 7) Specific mutations </h4>
<p> To be completed. </p>





<p> For further information, see the FAQ.</p>

<br>
<br>
<br>

{% endblock %}
