{% extends "base.html" %}

{% block content %}

<center>
  <br>
  <h3>Upload human SNPs in VCF format</h3>
  <br>

    <!-- modal body -->
    <div id="modal1" class="modal">
      <div class="modal-content">
        <h4>Example of an VCF file:</h4>
        <div style="text-align: center">
	  <pre>
          <div align="left" style="display: inline-block !important">
          #CHROM  POS ID  REF ALT 
          1   160092317   .   C   T 
          1   116967141   .   C   T 
          6   129250113   .   T   C 
          15  34359248    .   G   C 
          15  34359655    .   G   A 
          13  110642401   .   G   A 
          13  76992000    .   G   C 
          2   206786779   .   A   T 
          17  80108808    .   C   T 
          17  80118731    .   G   A 
          1   235806656   .   A   T 
          17  9023271     .   G   T 
          19  18166175    .   C   T 
          13  36819698    .   A   T 
          20  63690136    .   C   T 
          9   128607988   .   C   G 
          1   47280395    .   G   A 
          19  36058855    .   G   A 
          </div>
	  </pre>
        </div>
      </div>
      <div class="modal-footer">
        <a href="#!" class="modal-close waves-effect waves-green btn-flat">Close</a>
      </div>
    </div>


  <div class="row">
    <div class="col offset-s3 s6">
      <p style="text-align: justify"> 
        At the moment, <span style="font-variant:small-caps"> MutationExplorer </span> automatically loads AlphaFold models
        for human transcripts with 2700 or fewer residues.  GRCh38 coordinates are required in the VCF file, which is forwarded to ENSEMBL VEP to identify missense variant(s).  Only missense variants which impact canonical UniProtKB/Swiss-Prot transcripts are further explored. 
      </p>
      <p> Details of the workflow are outlined in FAQs. </p>
      <br>
      <div id="err" style="color:red;"><b>{{error}}</b></div>
    </div>
  </div>
  



</center>
<form id="submitForm" method="POST" enctype="multipart/form-data" class="row">

<div id="radio-field" class="input-field col offset-s3 s6">
  <span id="radio-title" class="grey-text text-darken-1">Minimization <a class="waves-effect waves-light btn modal-trigger grey lighten-2" href="#modalmin" title="Further information" style="float:right;">i</a></span>

      <!-- Modal Structure -->
      <div id="modalmin" class="modal">
	<div class="modal-content">
	  <h4>Minimization</h4>
	  <p>
	    The minimization is crucial for the quality of the outcome. The better the structure is energetically minimized, the more reliable the results will be. <br>
	    We continuously pre-minimize structures from the PDB using the commandline described below.
	    If you specify a PDB ID that is contained in our database, the pre-minimized structure is used. <br>
	    Otherwise we offer two modes of side-chain minimization, a short and a long one. The short one should take a few minutes, depending on the size of the protein. The long one may take some hours for large proteins. 
	  </p>
	  <h4> AlphaFold and your own models</h4>
	  <p>These are not pre-minimized and you should follow the steps below. The AlphaFold models should be minimized, but the minimization has to be performed with the same energy functions than we use in the <span class="mut-ex">MutationExplorer</span>, namely Rosetta. </p>
	  
	  <h4> Do it yourself!</h4>
	  <p> The minimizations we offer as 'long' and 'short' are both only side-chain optimizations.
	    However, we recommend you to perform the following minimization. It is optimizing also the backbone in a limited way (due to '-relax:constrain_relax_to_start_coords'). A free backbone minimization might be leading to problems for e.g. membrane proteins. </p>
	  <div class="codebox">
	    <code>
	      PATH_TO_ROSETTA/relax.static.linuxgccrelease   -relax:fast true -relax:cartesian true -score:weights ref2015_cart -use_input_sc -optimization::default_max_cycles 200 -linmem_ig 10 -relax:constrain_relax_to_start_coords -ex1 -ex2  -nstruct 20 -in:file:s  INPUT.pdb -out:pdb -out:prefix  OUTDIR/
	    </code>
	  </div>

	  <p>
	    This commandline will create 20 models, from which you should select the one with the lowest energy. Each PDB contains a line starting with 'pose'. In this line you want the last value, which is the total energy.</p>
	  
	</div>
	<div class="modal-footer">
	  <a href="#!" class="modal-close waves-effect waves-green btn-flat">Close</a>
	</div>
      </div>
  
  <p>
    <label>
      <input type="radio" name="min-selector" value="long" checked required />
      <span>long (5-60 min, recommended)</span>
    </label>
  </p>
  <p>
    <label>
      <input type="radio" name="min-selector" value="short" />
      <span>quick (1-3 min)</span>
    </label>
  </p>

</div>
<div id="radio-field" class="input-field col offset-s3 s6">
  <span id="radio-title" class="grey-text text-darken-1">RaSP Model <a class="waves-effect waves-light btn modal-trigger grey lighten-2" href="#modalrasp" title="Further information" style="float:right;">i</a></span>
  
  <!-- Modal Structure -->
      <div id="modalrasp" class="modal">
	<div class="modal-content">
	  <h4>RaSP model</h4>
	  <p>
	   RaSP is a deep-learning-based tool that rapidly estimates
protein stability changes. With RaSP, Mutation Explorer presents the user with a quick initial estimation of a mutation's (de)stabilizing effect, without their having to wait for the longer full minimization process.  The RaSP tool consists of two linearly linked networks.  A self-supervised 3D convolutional neural network that has learned representations of protein structures is followed by a fully connected neural network that maps these internal representations to Rosetta protein stability changes. RaSP is optimised for accuracy in the range [-1,7] kcal/mol.	   
	   </p>
	    <p>
	    This command will create a RaSP model for each chain of the structure. This can take up to 10 minutes per chain.</p>
  	</div>
	<div class="modal-footer">
	  <a href="#!" class="modal-close waves-effect waves-green btn-flat">Close</a>
	</div>
      </div>
  
     <p>
    <label>
      <input type="checkbox" name="rasp-checkbox"  checked />
      <span>Calculate RaSP model</span>
    </label>
  </p>
</div>


<form id="submitForm" method="POST" enctype="multipart/form-data" class="row">
  

  <!-- modal vcf demo 
  <div class="col offset-s2 s8">
    <button data-target="modal1" class="btn-small modal-trigger grey lighten-1" title="Further information" style="float: right;">demo vcf file</button>
  </div>


<br>
<br>
<br>-->

<div class="col offset-s3 s6 center-align">
  <div class="">
    <div style="width: 100%; display: flex; justify-content: center">
      <div class="file-field input-field" style="width: 100%">
	<div class="btn" style="width: 100%">
	  <span>Upload a VCF file</span>
	  <input type="file" name="vcf" id="vcf" accept=".vcf">
	</div>
      </div>
    </div>
  </div>
</div>


<br>

<div class="col offset-s3 s6 center-align">
    <span id="error-msg">{{error}}</span>
</div>

<div class="col offset-s3 s6 center-align">
<input id="submit-btn" type="submit" value="submit" class="btn">
<br>
<br>
<br>
<div style="display: flex; justify-content: center">
<a href="{{url_for('load_example', tag='vcf_demo')}}" class="btn-small grey lighten-1">load example</a>
<div style="width: 2vh"></div>
<button data-target="modal1" class="btn-small modal-trigger grey lighten-1" title="Further information">demo vcf file</button>
</div>
</div>



</form>

<br>
<br>



{% endblock %}


{% block scripts %}
<script>

    function showFilename() {
        var fname = $(this).val().replace(/^.*[\\\/]/, '')
        $(this).parent().parent().find(".filename").html(fname)
    }

    function initUploads() {
        var uploads = $(".file-field")
        for (i = 0; i < uploads.length; i++) {
            $(uploads[i]).append('<center><div class="filename"></div></center>')
            $($(uploads[i]).find("input")).on("change", showFilename)
        }
    }

    initUploads()

</script>
{% endblock %}
